servicenow/stackstate_checks/servicenow/servicenow.py

```diff
@@ -131,37 +132,42 @@ def _collect_components(self, instance_config, timeout):
-        return self._get_json(url, timeout, auth)
+        return self._get_json_batch(url, offset, batch_size, timeout, auth)
 
-    def _process_components(self, instance_config, timeout):
+    def _process_components(self, instance_config, batch_size, timeout):
         """
         process components fetched from CMDB
         :return: nothing
         """
         instance_tags = instance_config.instance_tags
+        offset = 0
 
-        state = self._collect_components(instance_config, timeout)
-
-        if "result" in state:
-            for component in state.get('result', []):
-                data = {}
-                component = self.filter_empty_metadata(component)
-                identifiers = []
-                comp_name = component.get('name')
-                comp_type = component.get('sys_class_name')
-                external_id = component.get('sys_id')
-
-                if 'fqdn' in component and component['fqdn']:
-                    identifiers.append("urn:host:/{}".format(component['fqdn']))
-                if 'host_name' in component and component['host_name']:
-                    identifiers.append("urn:host:/{}".format(component['host_name']))
-                else:
-                    identifiers.append("urn:host:/{}".format(comp_name))
-                identifiers.append(external_id)
-                data.update(component)
-                data.update({"identifiers": identifiers, "tags": instance_tags})
-
-                self.component(external_id, comp_type, data)
+        completed = False
+        while not completed:
+            components = self._collect_components(instance_config, offset, batch_size, timeout)
+            total_components = len(components.get("result"))
+            if "result" in components and isinstance(components["result"], list):
+                for component in components.get('result'):
+                    data = {}
+                    component = self.filter_empty_metadata(component)
+                    identifiers = []
+                    comp_name = component.get('name')
+                    comp_type = component.get('sys_class_name')
+                    external_id = component.get('sys_id')
+
+                    if 'fqdn' in component and component['fqdn']:
+                        identifiers.append("urn:host:/{}".format(component['fqdn']))
+                    if 'host_name' in component and component['host_name']:
+                        identifiers.append("urn:host:/{}".format(component['host_name']))
+                    else:
+                        identifiers.append("urn:host:/{}".format(comp_name))
+                    identifiers.append(external_id)
+                    data.update(component)
+                    data.update({"identifiers": identifiers, "tags": instance_tags})
+
+                    self.component(external_id, comp_type, data)
+            completed = total_components < batch_size
```

Comment:
1. `total_components` is renamed to `number_of_components_in_current_batch`
2.  I didn't find a way to get total of components in Servicenow table api

Reply:
Regarding question 2. I found one header that can be useful.
![Screenshot 2020-10-16 at 18 28 15](https://user-images.githubusercontent.com/39823032/96284095-6ca6ec80-0fdd-11eb-80fb-ee334e42493b.png)

I'm not sure from docs if `X-Total-Count` will return real total value or max batch_size 10000.

I believe this can be tested at client environment if need batching. We can add info log with this value, and do refactoring if the `X-Total-Count` is correct.

Reply:
@hruhek this `X-Total-Count` refers to the total count for that API endpoint and will not give default value.

Reply:
@vishalxebia even when result is more than max batch_size 10000?